test/schema_test.exs

```diff
@@ -138,4 +146,60 @@ defmodule OpenApiSpex.SchemaTest do
     assert :ok = Schema.validate(schema, "bar", %{})
   end
 
+  test "Validate schema type object when value is array" do
+    schema = %Schema{
+      type: :object
+    }
+    assert {:error, _} = Schema.validate(schema, [], %{})
+  end
+
+  test "Validate schema type array when value is object" do
+    schema = %Schema{
+      type: :array
+    }
+    assert {:error, _} = Schema.validate(schema, %{}, %{})
+  end
+
+  test "Validate schema type boolean when value is object" do
+    schema = %Schema{
+      type: :boolean
+    }
+    assert {:error, _} = Schema.validate(schema, %{}, %{})
+  end
+
+  test "Validate schema type string when value is object" do
+    schema = %Schema{
+      type: :string
+    }
+    assert {:error, _} = Schema.validate(schema, %{}, %{})
+  end
+
+  test "Validate schema type string when value is DateTime" do
+    schema = %Schema{
+      type: :string
+    }
+    assert :ok = Schema.validate(schema, DateTime.utc_now(), %{})
+  end
+
+  test "Validate schema type object when value is DateTime" do
+    schema = %Schema{
+      type: :object
+    }
+    assert :ok = Schema.validate(schema, DateTime.utc_now(), %{})
```

Comment:
Should this be valid? I would expect DateTimes to serialize to / from strings only.

Reply:
Yes, I think we should add a case for the value being a `DateTime`, and non-string schema types.
Something along the lines of:

```
def validate(schema = %Schema{type: :string}, value, path, _schemas) when is_binary(value) do
 ...
end
def validate(schema = %Schema{type: :string, format: :"date-time"}, value = %DateTime{}, path, _schemas) do
  :ok
end
def validate(schema = %Schema{}, value = %DateTime{}, path, _schemas) do
  {:error, "#{path}: DateTime can only be validated against type: :string, format: date-time"}
end
```